feat: adds VerticalBlockRenderCompleted filter

tecoholic opened the pull request:
**Description:** 
This adds `VerticalBlockRenderCompleted` filter that will be passed the rendered fragment of a vertical block, along with the context and the view in which it was rendered. 

This will allow for changing the rendered content by the pipeline steps. 

**JIRA:** NA

**Dependencies:** 
- https://github.com/openedx/edx-platform/pull/31388

**Merge deadline:**  None

**Installation instructions:** Nothing special

**Testing instructions:**

Detailed testing instructions with a sample pipeline step are available in the edx-paltform PR: https://github.com/openedx/edx-platform/pull/31388

**Reviewers:**
- [ ] @navinkarkera 
- [ ]  @mariajgrimaldi 

**Merge checklist:**
- [ ] All reviewers approved
- [ ] CI build is green
- [ ] Version bumped
- [ ] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [ ] Commits are squashed

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPI after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:**

Nothing comes to mind.

openedx-webhooks commented:
Thanks for the pull request, @tecoholic!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

<!-- comment:end_of_wip -->




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IHRydWV9 -->

navinkarkera commented:
@tecoholic  Nice work! :+1:
- [x] I tested this: Tested the filter as described in https://github.com/openedx/edx-platform/pull/31388
- [x] I read through the code
- [ ] I checked for accessibility issues
- [x] Includes documentation
- [ ] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

mphilbrick211 commented:
Flagging this for review @mariajgrimaldi :)

mariajgrimaldi commented:
I'm back on monday and I'll take a look at this. Sorry for the delay!

mariajgrimaldi commented on openedx_filters/learning/filters.py:
```diff
@@ -442,3 +442,25 @@ def run_filter(cls, block, context):
         return data.get("block"), data.get("context")
+
+
+class VerticalBlockRenderCompleted(OpenEdxPublicFilter):
```
Something I just noticed is both VerticalBlockChildRenderStarted and this filter are missing the exceptions like the ones defined in eg. [here](https://github.com/openedx/openedx-filters/pull/52/files#diff-0c79edd111d62405dc0829ac0b0b971c73b410de87c0133a7190b3de08b40ed9R354). Is there a reason for it?

mariajgrimaldi replied:
Then, when you implement those exceptions, you'll need to catch them in the edx-platform PR. You could follow [this implementation](https://github.com/openedx/edx-platform/blob/9a0fbe1a5758e80cd378b52bc93ef245ad458069/common/djangoapps/student/views/dashboard.py#L897) as a guide. You don't have to implement all exceptions, just the ones that make sense.

mariajgrimaldi reviewed (changes requested):
I left a few comments regarding the missing exceptions. I missed those while reviewing https://github.com/openedx/openedx-filters/pull/38, so maybe we'll need to revisit that implementation as well. Let me know what you think.

mphilbrick211 commented:
Hi @tecoholic! Just following up on the changes Maria requested.

tecoholic commented:
@mphilbrick211 Hi, I was a bit busy last couple of weeks. Kindly expect updates to this PR this week. :)

tecoholic commented:
@mariajgrimaldi Hi, I have introduced exceptions for both the `VerticalBlockRenderCompleted` filter and the previous `VerticalBlockChildRenderStarted` filter and updated the PR. Can you kindly review this again?

mariajgrimaldi commented on openedx_filters/learning/filters.py:
```diff
@@ -466,3 +471,30 @@ def run_filter(cls, enrollments):
+
+    filter_type = "org.openedx.learning.vertical_block.render.completed.v1"
+
+    class PreventVerticalBlockRender(OpenEdxFilterException):
```
I have a few questions here, but first I want to apologize for the missing documentation. I'm working on it, but it's been difficult to find time for it.

Alright, here are the questions:
1. Is this a filter that interacts with templates?
2. If that's true, which exceptions make sense to use? We've been using these:
- RenderInvalidComponent where component could be, eg. RenderInvalidDashboard: renders an error message, eg. you can't view this component
- RedirectToPage: redirects to a different page. Maybe you want to render a page hosted elsewhere. I don't think that one makes sense here. What do you think?
- RenderCustomResponse: render any other response. Like 404.

Which of those do you think would apply here?

mariajgrimaldi commented:
I tested this using the edx-platform PR and it works nicely! Thank you. @navinkarkera do you have any comments on this?

navinkarkera commented:
@mariajgrimaldi @tecoholic Looks good to me. We also have a sandbox demo which displays a form using this filter.

Sandbox:
- [URL](https://app.skilltagging.opencraft.hosting/learning/course/course-v1:edX+DemoX+Demo_Course/block-v1:edX+DemoX+Demo_Course+type@sequential+block@19a30717eff543078a5d94ae9d6c18a5/block-v1:edX+DemoX+Demo_Course+type@vertical+block@4f6c1b4e316a419ab5b6bf30e6c708e9)
- Login using staff credentials.
- Scroll to end of the page.

tecoholic commented on openedx_filters/learning/filters.py:
```diff
@@ -431,6 +431,11 @@ class VerticalBlockChildRenderStarted(OpenEdxPublicFilter):
 
     filter_type = "org.openedx.learning.vertical_block_child.render.started.v1"
 
+    class PreventChildBlockRender(OpenEdxFilterException):
+        """
+        Custom class used to stop a particular child block from being rendered.
+        """
```
@mariajgrimaldi I don't have any specific issues about moving this to a particular PR. But this is a part of the broader Skills Tagging project and I am not sure how project management would be affected because of it.

@navinkarkera Would creating a new ticket/PR for this be possible? Or should we use the same ticket and add an extra PR to it?

> There are others missing still,

Also, can you kindly elaborate on this? Is there a specific set of exceptions that must be present for all the filters?  From what I observed from the other filters, the exceptions seem to be very context specific. In this case of the `VerticalBlockChildRenderStarted` filter, I couldn't come up with any more scenarios that could be implemented by the filters. For example, an exception like "ReplaceBlock" won't be possible as that would require the filter to have access to the ModuleSystem that's rendering the XBlock and a lot more context that the filter won't have.

Similarly, for a the `VeriticalBlockRenderCompleted` hook, any changes that need to be done can only done to the output HTML Fragment and the reasonable exception scenario seems prevent the entire block, but other than that, I am finding it difficult to come up with any more exception scenarios.

navinkarkera replied:
@tecoholic We can use the same ticket and add the PR to it.

mariajgrimaldi replied:
You're right about the exceptions! @tecoholic 
This PR can stay as it is. Thanks for talking through the suggestions!

mariajgrimaldi reviewed (approved):
@tecoholic @navinkarkera: this looks awesome! thanks for your hard work!
![image](https://user-images.githubusercontent.com/64440265/219046839-84f7a1ba-8d28-41a4-8c7a-36f9b749ce4e.png)

mariajgrimaldi commented:
@tecoholic @navinkarkera: is this ready for merging?

mariajgrimaldi commented:
@tecoholic: just one last thing, can you please update the changelog? Thanks!

tecoholic commented:
@mariajgrimaldi  Thank you for your time and the review. I have added the changes to the changelog and bumped the version. The CI seems to be failing due to a [Codecov token issue](https://github.com/openedx/openedx-filters/actions/runs/4195061648/jobs/7274062381#step:7:53). Other than that I think we are all good. Kindly merge and release at your convenience. :tada:

mariajgrimaldi commented:
@tecoholic: hi again! can you rebase with master? I merged a PR updating the codecov version!

tecoholic commented:
@mariajgrimaldi Done. And all checks seem to be passing.

openedx-webhooks commented:
@tecoholic 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-filters&entry.867055334=https://github.com/openedx/openedx-filters/pull/52&entry.1484655318=https://github.com/tecoholic&entry.752974735=2022-12-02+09:45&entry.1917517419=2023-02-17+13:45&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

mariajgrimaldi commented:
@tecoholic: new release is out!
